sound/soc/sdw_utils/soc_sdw_rt_mf_sdca.c

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// This file incorporates work covered by the following copyright notice:
+// Copyright (c) 2024 Intel Corporation.
+
+/*
+ *  soc_sdw_rt_mf_sdca - Helpers to handle RT712/RT722 from generic machine driver
+ */
+
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include <sound/control.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include <sound/soc-dapm.h>
+#include <sound/soc_sdw_utils.h>
+
+#define CODEC_NAME_SIZE	7
+
+/* dapm routes for rt712/rt722 will be registered dynamically */
+static const struct snd_soc_dapm_route rt712_spk_map[] = {
+	{ "Speaker", NULL, "rt712 SPOL" },
+	{ "Speaker", NULL, "rt712 SPOR" },
+};
+
+static const struct snd_soc_dapm_route rt722_spk_map[] = {
+	{ "Speaker", NULL, "rt722 SPK" },
+};
+
+static const struct snd_soc_dapm_route *get_codec_name_and_route(struct snd_soc_dai *dai,
+								 char *codec_name)
+{
+	/* get the codec name */
+	snprintf(codec_name, CODEC_NAME_SIZE, "%s", dai->name);
+
+	/* choose the right codec's map  */
+	if (strcmp(codec_name, "rt712") == 0)
+		return rt712_spk_map;
+	else if (strcmp(codec_name, "rt722") == 0)
+		return rt722_spk_map;
+	else
+		return NULL;
+}
+
+int asoc_sdw_rt_mf_sdca_spk_rtd_init(struct snd_soc_pcm_runtime *rtd, struct snd_soc_dai *dai)
+{
+	struct snd_soc_card *card = rtd->card;
+	const struct snd_soc_dapm_route *rt_mf_sdca_map;
+	char codec_name[CODEC_NAME_SIZE];
+	struct snd_soc_dai *codec_dai;
+	int ret;
+	int i;
+
+	rt_mf_sdca_map = get_codec_name_and_route(dai, codec_name);
+
+	card->components = devm_kasprintf(card->dev, GFP_KERNEL,
+					  "%s spk:%s",
+					  card->components, codec_name);
+	if (!card->components)
+		return -ENOMEM;
+
+	for_each_rtd_codec_dais(rtd, i, codec_dai) {
+		if (strstr(codec_dai->component->name_prefix, "-1"))
+			ret = snd_soc_dapm_add_routes(&card->dapm, rt_mf_sdca_map, 2);
+		else if (strstr(codec_dai->component->name_prefix, "-2"))
+			ret = snd_soc_dapm_add_routes(&card->dapm, rt_mf_sdca_map + 2, 2);
+	}
```

Comment:
I don't think there will be "-1" or "-2" for multi function speakers. Just add `rt_mf_sdca_map` with `ARRAY_SIZE(rt_mf_sdca_map)`